### PR TITLE
Adding the ability to stop staking via api

### PR DESCRIPTION
**Pull request by dangershony**

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs [hunk 1]
```diff
@@ -114,5 +113,32 @@ public IActionResult StartStaking([FromBody]StartStakingRequest request)
...
+        /// Stop staking.
+        /// </summary>
+        /// <returns>An OKResult object that produces a status code 200 HTTP response.</returns>
+        [Route("stopstaking")]
+        [HttpPost]
+        public IActionResult StopStaking()
+        {
+            try
+            {
+                if (!this.ModelState.IsValid)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs]: Do you need this if there is no model?

File: src/Stratis.Bitcoin.Features.Miner/PosMinting.cs [hunk 2]
```diff
@@ -402,14 +408,37 @@ public IAsyncLoop Stake(WalletSecret walletSecret)
...
+            if (this.stakingLoop == null)
+            {
+                this.logger.LogTrace("(-)[NOT_MINING]");
+                return;
+            }
+
+            this.stakeCancellationTokenSource.Cancel();
+            this.stakingLoop.Dispose();
+            this.stakingLoop = null;
+            this.stakeCancellationTokenSource = null;
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Features.Miner/PosMinting.cs]: how about disposing first?

File: src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs (hunk 1 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs]: You tell me, I am not sure.

File: src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs [hunk 3]
```diff
@@ -114,5 +113,32 @@ public IActionResult StartStaking([FromBody]StartStakingRequest request)
                 return ErrorHelpers.BuildErrorResponse(HttpStatusCode.BadRequest, e.Message, e.ToString());
             }
         }
+
+        /// <summary>
+        /// Stop staking.
+        /// </summary>
+        /// <returns>An OKResult object that produces a status code 200 HTTP response.</returns>
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs]: OKResult?

File: src/Stratis.Bitcoin.Features.Miner/PosMinting.cs (hunk 2 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Miner/PosMinting.cs]: Disposing blocks

File: src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs (hunk 3 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs]: That's the name of the type, I can add a ref=

**Comment by bokobza** [comment]
I'm ok with this, @Aprogiena do you want to have a look?

File: src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs]: Yes, you don't need to check the ModelState here as you didn't post anything to the controller method.

Comment (Neurosploit) [reply src/Stratis.Bitcoin.Features.Miner/Controllers/MinerController.cs]: Yes I think the modelstate.isvalid can be removed on any controller method that does not have any parameters. I think you can remove it for GetStakingInfo() as well.

File: src/Stratis.Bitcoin.Features.Miner/PosMinting.cs (hunk 2 above)
Comment (Aprogiena) [reply src/Stratis.Bitcoin.Features.Miner/PosMinting.cs]: what?

Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Miner/PosMinting.cs]: Here: https://docs.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads it says:

> The CancellationTokenSource class implements the IDisposable interface. You should be sure to call the CancellationTokenSource.Dispose method when you have finished using the cancellation token source to free any unmanaged resources it holds.

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Miner/PosMinting.cs]: No but I mean the dispose implementation of the async loop blocks until the loop ends.
So we have to cancel first then wait for the loop to finish calling the Dispose()
Then we can dispose the cancellation.

**Pull request merged**